File: quant-black-box/backend/app/routers/apt.py
from fastapi import APIRouter, HTTPException

from .. import engine
from ..schemas import AptParams as AptParamsModel, AptRequest, AptResponse, GridPoint, GridResponse

router = APIRouter()


def build_grid(req: AptRequest) -> GridResponse:
    g = req.grid
    n = g.n
    params = engine.AptParams(
        r=req.r, lam=req.lam, lams=req.lams, lamv=req.lamv, b3=req.b3, al=req.al
    )
    points: list[GridPoint] = []
    for i in range(n):
        x = g.b1_min + (g.b1_max - g.b1_min) * i / (n - 1)
        for j in range(n):
            y = g.b2_min + (g.b2_max - g.b2_min) * j / (n - 1)
            z = engine.apt_ret(x, y, g.metric == "alpha", params)
            points.append(GridPoint(x=x, y=y, z=z))
    return GridResponse(
        n=n,
        metric=g.metric,
        points=points,
        shape=[n, n],
        x_range=[g.b1_min, g.b1_max],
        y_range=[g.b2_min, g.b2_max],
    )


@router.post("", response_model=AptResponse)
def price(req: AptRequest) -> AptResponse:
    if req.grid is not None:
        if req.grid.b1_max <= req.grid.b1_min:
            raise HTTPException(status_code=422, detail="grid.b1_max must be > grid.b1_min")
        if req.grid.b2_max <= req.grid.b2_min:
            raise HTTPException(status_code=422, detail="grid.b2_max must be > grid.b2_min")

    params = engine.AptParams(
        r=req.r, lam=req.lam, lams=req.lams, lamv=req.lamv, b3=req.b3, al=req.al
    )
    ret = engine.apt_ret(req.b1, req.b2, req.metric == "alpha", params)

    return AptResponse(
        params=AptParamsModel(
            r=req.r, lam=req.lam, lams=req.lams, lamv=req.lamv, b3=req.b3, al=req.al
        ),
        metric=req.metric,
        b1=req.b1,
        b2=req.b2,
        ret=ret,
        grid=build_grid(req) if req.grid is not None else None,
    )

File: quant-black-box/server/src/routes/apt.ts
import { Router } from 'express';
import { z } from 'zod';
import { aptRet } from '../engine.js';

const router = Router();

const aptGridSchema = z.object({
  n: z.coerce.number().min(2).max(120).default(30),
  b1_min: z.coerce.number().default(0),
  b1_max: z.coerce.number().default(2),
  b2_min: z.coerce.number().default(-1),
  b2_max: z.coerce.number().default(1),
  metric: z.enum(['fair', 'alpha']).default('fair'),
});

const aptSchema = z.object({
  r: z.coerce.number(),
  lam: z.coerce.number(),
  lams: z.coerce.number(),
  lamv: z.coerce.number(),
  b3: z.coerce.number(),
  al: z.coerce.number(),
  metric: z.enum(['fair', 'alpha']).default('fair'),
  b1: z.coerce.number().default(1),
  b2: z.coerce.number().default(0.5),
  grid: aptGridSchema.nullish(),
});

router.post('/', (req, res) => {
  const parsed = aptSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(422).json({ ok: false, error: 'Invalid parameters', details: parsed.error.flatten() });
    return;
  }
  const { r, lam, lams, lamv, b3, al, metric, b1, b2, grid } = parsed.data;
  const params = { r, lam, lams, lamv, b3, al };
  const ret = aptRet(b1, b2, metric === 'alpha', params);

  let gridData = null;
  if (grid) {
    if (grid.b1_max <= grid.b1_min) {
      res.status(422).json({ ok: false, error: 'grid.b1_max must be > grid.b1_min' });
      return;
    }
    if (grid.b2_max <= grid.b2_min) {
      res.status(422).json({ ok: false, error: 'grid.b2_max must be > grid.b2_min' });
      return;
    }
    const n = grid.n;
    const points: { x: number; y: number; z: number }[] = [];
    for (let i = 0; i < n; i++) {
      const x = grid.b1_min + (grid.b1_max - grid.b1_min) * i / (n - 1);
      for (let j = 0; j < n; j++) {
        const y = grid.b2_min + (grid.b2_max - grid.b2_min) * j / (n - 1);
        points.push({ x, y, z: aptRet(x, y, grid.metric === 'alpha', params) });
      }
    }
    gridData = { n, metric: grid.metric, points, shape: [n, n], x_range: [grid.b1_min, grid.b1_max], y_range: [grid.b2_min, grid.b2_max] };
  }

  res.json({ params: { r, lam, lams, lamv, b3, al }, metric, b1, b2, ret, grid: gridData });
});

export default router;
